yarn-project/pxe/src/storage/tagging_data_provider/sender_tagging_data_provider.ts

```diff
@@ -0,0 +1,210 @@
+   */
+  getLastFinalizedIndex(secret: DirectionalAppTaggingSecret): Promise<number | undefined> {
+    return this.#lastFinalizedIndexes.getAsync(secret.toString());
+  }
+
+  /**
+   * Returns the last used index for a given directional app tagging secret, considering both finalized and pending
+   * indexes.
+   * @param secret - The directional app tagging secret to query the last used index for.
+   * @returns The last used index.
+   */
+  async getLastUsedIndex(secret: DirectionalAppTaggingSecret): Promise<number | undefined> {
+    const lastFinalizedIndex = await this.#lastFinalizedIndexes.getAsync(secret.toString());
+    const pendingTxScopedIndexes = (await this.#pendingIndexes.getAsync(secret.toString())) ?? [];
+    const pendingIndexes = pendingTxScopedIndexes.map(entry => entry.index);
+
+    if (pendingTxScopedIndexes.length === 0) {
+      return lastFinalizedIndex;
+    }
+
+    // As the last used index we return the highest one from the pending indexes. Note that this might not technically
+    // be the last used index in case we sent logs from multiple PXEs in parallel but that's just a detail.
+    return Math.max(...pendingIndexes);
+  }
+
+  /**
+   * A tx that contained private logs with tags corresponding to some of the indexes returned from this data provider
+   * has been dropped so we delete the corresponding pending indexes. This results in the indexes being reused which
+   * results in the txs that reused the indexes to be linked with the dropped tx. We don't worry about that for now.
+   * @param txHash - The hash of the tx to drop the pending indexes for.
+   */
+  async dropPendingIndexes(txHash: TxHash) {
+    const txHashStr = txHash.toString();
+    const allSecrets = await toArray(this.#pendingIndexes.keysAsync());
+
+    for (const secret of allSecrets) {
+      const pendingData = await this.#pendingIndexes.getAsync(secret);
+      if (pendingData) {
+        const filtered = pendingData.filter(item => item.txHash.toString() !== txHashStr);
+        if (filtered.length === 0) {
+          await this.#pendingIndexes.delete(secret);
+        } else if (filtered.length !== pendingData.length) {
+          // Some items were filtered out, so update the pending data
+          await this.#pendingIndexes.set(secret, filtered);
+        }
+        // else: No items were filtered out (txHash not found for this secret) --> no-op
+      }
+    }
+  }
+
+  /**
+   * A tx that contained private logs with tags corresponding to some of the indexes returned from this data provider
+   * has been finalized so we update the corresponding pending indexes to be finalized (we move them from pending map
+   * to finalized map).
+   * @param txHash - The hash of the tx to update the status of.
+   */
+  async updateStatusToFinalized(txHash: TxHash) {
+    const txHashStr = txHash.toString();
+    const allSecrets = await toArray(this.#pendingIndexes.keysAsync());
+
+    for (const secret of allSecrets) {
```

Comment:
There's 4 levels of conditional nesting here, I'd try to find a simpler structure to express these ideas. e.g.

```
for secret of secrets {
  let pending = pending.get(secret)
  if !pending { cotninue }
  
  let matching = pending.filter(i.tx == tx).map(i.index);
  if matching.len == 0 { continue }
  else if matching.len > 1 {
    throw new error("this should never happen");
  }
  
  // updates
}
```